src/Contracts/Pools/OpdexStakingPool.cs

```diff
@@ -59,105 +54,137 @@ public UInt256 StakingRewardsBalance
+
         MintStakingRewards(ReserveCrs, ReserveSrc);
         
-        var stakedBalance = GetStakedBalance(Message.Sender);
-        if (stakedBalance > 0)
-        {
-            CollectStakingRewardsExecute(Message.Sender, stakedBalance, false);
-            UnstakeExecute(Message.Sender, stakedBalance, false);
-        }
+        UpdateStakingPosition(Message.Sender);
+
+        TotalStaked += amount;
         
-        stakedBalance += amount;
+        SetStakedBalance(Message.Sender, GetStakedBalance(Message.Sender) + amount);
         
         SafeTransferFrom(StakingToken, Message.Sender, Address, amount);
         
-        SetStakedBalance(Message.Sender, stakedBalance);
-        SetStakingWeightExecute(stakedBalance);
+        Log(new StakeLog { Staker = Message.Sender, Amount = amount, TotalStaked = TotalStaked, EventType = (byte)StakeEventType.StartStaking});
         
         NominateLiquidityPool();
         
+        UpdateKLast();
+        
         Unlock();
     }
     
     /// <inheritdoc />
-    public void Collect(Address to, bool liquidate)
+    public void CollectStakingRewards(bool liquidate)
     {
         EnsureUnlocked();
         EnsureStakingEnabled();
         
         MintStakingRewards(ReserveCrs, ReserveSrc);
         
-        var stakedBalance = GetStakedBalance(Message.Sender);
+        UpdateStakingPosition(Message.Sender);
         
-        CollectStakingRewardsExecute(to, stakedBalance, liquidate);
+        CollectStakingRewardsExecute(Message.Sender, liquidate);
         
-        SetStakingWeightExecute(stakedBalance);
+        UpdateKLast();
         
         Unlock();
     }
         
     /// <inheritdoc />
-    public void Unstake(Address to, bool liquidate)
+    public void StopStaking(UInt256 amount, bool liquidate)
     {
         EnsureUnlocked();
         EnsureStakingEnabled();
         
         MintStakingRewards(ReserveCrs, ReserveSrc);
         
+        UpdateStakingPosition(Message.Sender);
+        
         var stakedBalance = GetStakedBalance(Message.Sender);
         
-        CollectStakingRewardsExecute(to, stakedBalance, liquidate);
+        Assert(amount <= stakedBalance && amount > 0, "OPDEX: INVALID_AMOUNT");
+
+        TotalStaked -= amount;
         
-        UnstakeExecute(to, stakedBalance, true);
+        SetStakedBalance(Message.Sender, stakedBalance - amount);
+        
+        CollectStakingRewardsExecute(Message.Sender, liquidate);
+        
+        SafeTransferTo(StakingToken, Message.Sender, amount);
+        
+        Log(new StakeLog {Amount = amount, Staker = Message.Sender, TotalStaked = TotalStaked, EventType = (byte)StakeEventType.StopStaking});
         
         NominateLiquidityPool();
         
+        UpdateKLast();
+        
         Unlock();
     }
         
     /// <inheritdoc />
     public override UInt256 Mint(Address to)
     {
         EnsureUnlocked();
-        
+
         MintStakingRewards(ReserveCrs, ReserveSrc);
```

Comment:
Consider possible gas savings by assigning ReserveCrs & ReserveSrc to local variables again. Sacrifices brevity, but I'd say it's probably worth it.